File: sdk/node/src/resources/providers.ts
import { HttpClient } from '../http';
import { Provider } from '../types';

function parseProvider(data: Record<string, unknown>): Provider {
  const vramMib = (data.vram_mib as number) ?? 0;
  return {
    id: (data.id as number) ?? 0,
    name: (data.name as string) ?? '',
    gpuModel: (data.gpu_model as string) ?? '',
    vramMib,
    vramGb: Math.round((vramMib / 1024) * 10) / 10,
    status: (data.status as Provider['status']) ?? 'offline',
    reliabilityScore: (data.reliability_score as number) ?? 0,
  };
}

export class ProvidersResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * List all online GPU providers available for job submission.
   */
  async list(): Promise<Provider[]> {
    const data = await this.http.get<Record<string, unknown> | unknown[]>('/api/renters/available-providers');
    const raw = Array.isArray(data) ? data : ((data as Record<string, unknown>).providers as unknown[]) ?? [];
    return (raw as Record<string, unknown>[]).map(parseProvider);
  }

  /**
   * Fetch a single provider by ID.
   */
  async get(providerId: number): Promise<Provider> {
    const data = await this.http.get<Record<string, unknown>>(`/api/providers/${providerId}`);
    return parseProvider(data);
  }
}

File: sdk/python/dc1/resources/providers.py
"""Providers resource."""
from __future__ import annotations
from ..models import Provider
from ..exceptions import APIError


class ProvidersResource:
    def __init__(self, http):
        self._http = http

    def list(self) -> list[Provider]:
        """List all online GPU providers available for job submission.

        Returns:
            List of Provider objects with GPU specs and reliability scores.
        """
        data = self._http.get('/api/renters/available-providers')
        raw_list = data if isinstance(data, list) else data.get('providers', [])
        return [Provider.from_api(p) for p in raw_list]

    def get(self, provider_id: int) -> Provider:
        """Fetch a single provider by ID.

        Args:
            provider_id: The provider's numeric ID.

        Returns:
            Provider object.
        """
        providers = self.list()
        for provider in providers:
            if provider.id == provider_id:
                return provider
        raise APIError(
            f'Provider {provider_id} not found in available providers',
            status_code=404,
            response={'provider_id': provider_id},
        )
